aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,354 @@
+/// A smart table implementation based on linear hashing. (https://en.wikipedia.org/wiki/Linear_hashing)
+/// Compare to Table, it uses less storage slots but has higher chance of collision, it's a trade-off between space and time.
+/// Compare to other implementation, linear hashing splits one bucket a time instead of doubling buckets when expanding to avoid unexpected gas cost.
+/// SmartTable uses faster hash function SipHash instead of cryptographically secure hash functions like sha3-256 since it tolerates collisions.
+module aptos_std::smart_table {
+    use std::error;
+    use std::vector;
+    use aptos_std::aptos_hash::sip_hash_from_value;
+    use aptos_std::table_with_length::{Self, TableWithLength};
+    use aptos_std::type_info::size_of_val;
+    use aptos_std::math64::max;
+
+    /// Key not found in the smart table
+    const ENOT_FOUND: u64 = 1;
+    /// Smart table capacity must be larger than 0
+    const EZERO_CAPACITY: u64 = 2;
+    /// Cannot destroy non-empty hashmap
+    const ENOT_EMPTY: u64 = 3;
+    /// Key already exists
+    const EALREADY_EXIST: u64 = 4;
+    /// Invalid load threshold percent to trigger split.
+    const EINVALID_LOAD_THRESHOLD_PERCENT: u64 = 5;
+    /// Invalid target bucket size.
+    const EINVALID_TARGET_BUCKET_SIZE: u64 = 6;
+
+    /// SmartTable entry contains both the key and value.
+    struct Entry<K, V> has copy, drop, store {
+        hash: u64,
+        key: K,
+        value: V,
+    }
+
+    struct SmartTable<K, V> has store {
+        buckets: TableWithLength<u64, vector<Entry<K, V>>>,
```

Comment:
Perhaps an idea, K,V are generic types and likely strings or so. Right now splitting a bucket means moving a (K, V) pair from one bucket into separate bucket. 

I can imagine something like a simple linear vector of K,V pairs where you just append to, with a hash table storing an index into the array. This would mean that splitting buckets just involves copying integer indices.

An additional benefit could be, let's say we use 32 bits for index to use an additional 32 bits as remainder of the hash value. This way we find the bucket based on the lower bits of the hash, and then we can compare the 32 high bits with the 32 high bits of the index. This way if we find a match we are almost completely sure we found the right element. Ie. we never have to compare K twice.

Reply:
The problem is that each bucket corresponds to one storage slot. The slots are not co-located contiguously in the same storage location. Moving items is then required when splitting buckets as they need to be moved from one storage slot to another.

Reply:
I'm not understanding how each things are actually modeled on the level of the blockchain. But it seems that currently splitting a bucket requires moving key value pairs, which can be prevented by using indices.

Reply:
the hashtable storing the index also needs modifications. moving kv pairs is actually modifications.